File: examples/nx_simple_controller.cpp
/**
 * @file nx_simple_controller.cpp
 * @brief Example Jetson NX program using the NxBridge API
 *
 * This demonstrates the simplest possible NX-side controller:
 *   1. Receives sensor data from RK3588
 *   2. Prints a summary once per second
 *   3. Sends a fixed standing-pose command (position hold)
 *
 * Build (on Jetson NX or x86 for testing):
 *   g++ -std=c++17 -I.. -o nx_simple_controller nx_simple_controller.cpp \
 *       ../nx_bridge/nx_bridge.cpp -lpthread
 *
 * Or via CMake: -DBUILD_NX_SDK=ON
 *
 * Usage:
 *   ./nx_simple_controller [robot_ip] [recv_port] [send_port]
 *
 * Defaults: robot_ip=127.0.0.1, recv_port=31001, send_port=31002
 */

#include "nx_bridge_api.hpp"

#include <iostream>
#include <iomanip>
#include <cstring>
#include <cmath>
#include <chrono>
#include <thread>

// ============================================================================
// Fixed standing pose for Lite3 (12 joints, 4 legs × 3 DOF each)
// ============================================================================
// Joint order: FL_HipX, FL_HipY, FL_Knee, FR_HipX, FR_HipY, FR_Knee,
//               HL_HipX, HL_HipY, HL_Knee, HR_HipX, HR_HipY, HR_Knee
//
// These are approximate standing-pose joint angles (radians).
// Adjusted for your specific robot if needed.

struct StandPose {
    float hip_x  =  0.0f;     // abduction — centered
    float hip_y  = -0.65f;    // hip pitch
    float knee   =  1.30f;    // knee
};

static const StandPose kStand = {0.0f, -0.65f, 1.30f};

// ============================================================================
// Example 1: Two-callback API (separate observation and action callbacks)
// ============================================================================

static NxJointCommand s_fixed_pose;   // pre-computed command to send every cycle
static int s_obs_count = 0;

void OnObservation(const NxRobotState& state) {
    s_obs_count++;

    // Print summary once per second (at ~50 Hz we get ~50 obs/s)
    if (s_obs_count % 50 == 0) {
        std::cout << "[Example] seq=" << state.seq
                  << " state=" << state.current_state
                  << " rpy=("
                  << std::fixed << std::setprecision(2)
                  << state.rpy[0] * 180.0 / M_PI << ", "
                  << state.rpy[1] * 180.0 / M_PI << ", "
                  << state.rpy[2] * 180.0 / M_PI << ") deg"
                  << " cmd_vel=(" << state.cmd_vel[0] << ", "
                  << state.cmd_vel[1] << ", " << state.cmd_vel[2] << ")"
                  << std::endl;
    }
}

NxJointCommand GetAction() {
    // Always return the same fixed standing pose.
    // RK3588 applies zero-order hold between NX updates.
    return s_fixed_pose;
}

void RunTwoCallbackExample(const std::string& robot_ip, int recv_port, int send_port) {
    std::cout << "\n=== Two-Callback Example ===" << std::endl;

    // Configure default PD gains (applied if the returned command has zeros)
    NxBridge::SetDefaultGains(30.0f, 1.0f);

    // Build the fixed standing-pose command
    std::memset(&s_fixed_pose, 0, sizeof(s_fixed_pose));
    const StandPose& sp = kStand;
    // FL: {hip_x, hip_y, knee}
    s_fixed_pose.joint_pos_des[0] = sp.hip_x;   s_fixed_pose.joint_pos_des[1] = sp.hip_y;   s_fixed_pose.joint_pos_des[2] = sp.knee;
    // FR: {-hip_x, hip_y, knee}  (mirror image)
    s_fixed_pose.joint_pos_des[3] = -sp.hip_x;  s_fixed_pose.joint_pos_des[4] = sp.hip_y;   s_fixed_pose.joint_pos_des[5] = sp.knee;
    // HL: {hip_x, hip_y, knee}
    s_fixed_pose.joint_pos_des[6] = sp.hip_x;   s_fixed_pose.joint_pos_des[7] = sp.hip_y;   s_fixed_pose.joint_pos_des[8] = sp.knee;
    // HR: {-hip_x, hip_y, knee}
    s_fixed_pose.joint_pos_des[9] = -sp.hip_x;  s_fixed_pose.joint_pos_des[10] = sp.hip_y;  s_fixed_pose.joint_pos_des[11] = sp.knee;

    // Set PD gains — these are the same for all joints
    for (int i = 0; i < 12; ++i) {
        s_fixed_pose.kp[i]     = 30.0f;
        s_fixed_pose.kd[i]     = 1.0f;
        s_fixed_pose.tau_ff[i] = 0.0f;
    }

    NxBridge bridge(recv_port, robot_ip, send_port);
    bridge.Start(OnObservation, GetAction, 20.0f);  // 50 Hz

    std::cout << "Running... Press Ctrl+C to stop." << std::endl;

    // Keep the main thread alive while the bridge runs in background threads
    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    bridge.Stop();
}

// ============================================================================
// Example 2: Single-callback API (simpler, for stateful policies)
// ============================================================================

void RunSingleCallbackExample(const std::string& robot_ip, int recv_port, int send_port) {
    std::cout << "\n=== Single-Callback Example ===" << std::endl;

    int tick = 0;

    // The single-callback version: function receives latest obs, returns action.
    auto policy = [&tick](const NxRobotState& state) -> NxJointCommand {
        tick++;
        NxJointCommand cmd;
        std::memset(&cmd, 0, sizeof(cmd));

        // Print every 50th tick
        if (tick % 50 == 0) {
            std::cout << "[Policy] tick=" << tick
                      << " rpy.x=" << state.rpy[0] * 180.0 / M_PI << " deg"
                      << std::endl;
        }

        // Fill with standing pose (same as two-callback example)
        const StandPose& sp = kStand;
        float poses[12] = {
            sp.hip_x, sp.hip_y, sp.knee,
            -sp.hip_x, sp.hip_y, sp.knee,
            sp.hip_x, sp.hip_y, sp.knee,
            -sp.hip_x, sp.hip_y, sp.knee
        };
        std::memcpy(cmd.joint_pos_des, poses, sizeof(poses));
        for (int i = 0; i < 12; ++i) {
            cmd.kp[i] = 30.0f;
            cmd.kd[i] = 1.0f;
        }
        return cmd;
    };

    NxBridge bridge(recv_port, robot_ip, send_port);
    bridge.Start(policy, 20.0f);  // 50 Hz

    std::cout << "Running... Press Ctrl+C to stop." << std::endl;

    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    bridge.Stop();
}

// ============================================================================
// Main
// ============================================================================

int main(int argc, char* argv[]) {
    std::string robot_ip = (argc > 1) ? argv[1] : "127.0.0.1";
    int recv_port = (argc > 2) ? std::atoi(argv[2]) : 31001;
    int send_port = (argc > 3) ? std::atoi(argv[3]) : 31002;

    std::cout << "=== Lite3 NX Simple Controller ===" << std::endl;
    std::cout << "Robot IP : " << robot_ip << std::endl;
    std::cout << "Recv port: " << recv_port << " (SensorPacket from RK3588)" << std::endl;
    std::cout << "Send port: " << send_port << " (CommandPacket to RK3588)" << std::endl;
    std::cout << "Stand pose: hip_x=" << kStand.hip_x
              << " hip_y=" << kStand.hip_y
              << " knee=" << kStand.knee << std::endl;

    // Run the two-callback example by default.
    // Change to RunSingleCallbackExample to test the simpler API.
    RunTwoCallbackExample(robot_ip, recv_port, send_port);

    return 0;
}

File: examples/nx_simple_controller.py
#!/usr/bin/env python3
"""
Example Jetson NX program using the NxBridge Python SDK.

Demonstrates two usage modes:
  1. Dual-callback: separate observation and action callbacks
  2. Single-callback: one function that receives obs and returns action

Usage:
    python nx_simple_controller.py [robot_ip] [recv_port] [send_port]

Defaults: robot_ip=127.0.0.1, recv_port=31001, send_port=31002
"""

import sys
import time
import math

# Add parent dir to path so we can import nx_bridge_py directly
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from nx_bridge_py import NxBridge, NxRobotState, NxJointCommand


# ============================================================================
# Fixed standing pose for Lite3 (12 joints, 4 legs × 3 DOF)
# ============================================================================
# Joint order: FL_HipX, FL_HipY, FL_Knee, FR_HipX, FR_HipY, FR_Knee,
#               HL_HipX, HL_HipY, HL_Knee, HR_HipX, HR_HipY, HR_Knee

STAND_POSE = {
    "hip_x":  0.00,
    "hip_y": -0.65,
    "knee":   1.30,
}


# ============================================================================
# Example 1: Dual-callback API
# ============================================================================

class DualCallbackExample:
    """Separate observation and action callbacks."""

    def __init__(self):
        self.obs_count = 0
        self.cmd = NxJointCommand()
        self.cmd.set_standing_pose(**STAND_POSE)

    def on_observation(self, state: NxRobotState):
        self.obs_count += 1
        if self.obs_count % 50 == 0:
            rpy_deg = tuple(v * 180.0 / math.pi for v in state.rpy)
            print(f"[Example] seq={state.seq} state={state.current_state} "
                  f"rpy=({rpy_deg[0]:.1f}, {rpy_deg[1]:.1f}, {rpy_deg[2]:.1f}) deg "
                  f"cmd_vel=({state.cmd_vel[0]:.2f}, {state.cmd_vel[1]:.2f}, {state.cmd_vel[2]:.2f})")

    def get_action(self) -> NxJointCommand:
        return self.cmd


def run_dual_callback(robot_ip: str, recv_port: int, send_port: int):
    print("\n=== Dual-Callback Example ===")
    example = DualCallbackExample()

    bridge = NxBridge(recv_port=recv_port, robot_ip=robot_ip, send_port=send_port)
    bridge.start(example.on_observation, example.get_action, period_ms=20)

    print("Running... Press Ctrl+C to stop.")
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\nShutting down...")
    finally:
        bridge.stop()


# ============================================================================
# Example 2: Single-callback API (simpler, for stateful policies)
# ============================================================================

def run_single_callback(robot_ip: str, recv_port: int, send_port: int):
    print("\n=== Single-Callback Example ===")
    tick = [0]  # use list for mutable closure

    def policy(state: NxRobotState) -> NxJointCommand:
        tick[0] += 1
        if tick[0] % 50 == 0:
            rpy_deg = state.rpy[0] * 180.0 / math.pi
            print(f"[Policy] tick={tick[0]} rpy.x={rpy_deg:.1f} deg")

        cmd = NxJointCommand()
        cmd.set_standing_pose(**STAND_POSE)
        return cmd

    bridge = NxBridge(recv_port=recv_port, robot_ip=robot_ip, send_port=send_port)
    bridge.start(policy, period_ms=20)

    print("Running... Press Ctrl+C to stop.")
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\nShutting down...")
    finally:
        bridge.stop()


# ============================================================================
# Example 3: Sinusoidal motion (demo of dynamic commands)
# ============================================================================

def run_sine_wave_example(robot_ip: str, recv_port: int, send_port: int):
    """Demonstrate a simple sinusoidal knee motion while holding standing pose."""
    print("\n=== Sine Wave Motion Example ===")
    start_time = time.time()

    def policy(state: NxRobotState) -> NxJointCommand:
        elapsed = time.time() - start_time
        # Small sinusoidal perturbation on knee joints (±0.1 rad at 0.5 Hz)
        knee_offset = 0.1 * math.sin(2.0 * math.pi * 0.5 * elapsed)

        cmd = NxJointCommand()
        sp = STAND_POSE
        poses = [
            sp["hip_x"],  sp["hip_y"],  sp["knee"] + knee_offset,   # FL
           -sp["hip_x"],  sp["hip_y"],  sp["knee"] + knee_offset,   # FR
            sp["hip_x"],  sp["hip_y"],  sp["knee"] + knee_offset,   # HL
           -sp["hip_x"],  sp["hip_y"],  sp["knee"] + knee_offset,   # HR
        ]
        cmd.joint_pos_des = list(poses)
        cmd.set_gains(kp=40.0, kd=1.5)  # slightly stiffer for dynamic motion
        return cmd

    bridge = NxBridge(recv_port=recv_port, robot_ip=robot_ip, send_port=send_port)
    bridge.start(policy, period_ms=20)

    print("Running sine wave motion... Press Ctrl+C to stop.")
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\nShutting down...")
    finally:
        bridge.send_shutdown()
        bridge.stop()


# ============================================================================
# Main
# ============================================================================

def main():
    robot_ip  = sys.argv[1] if len(sys.argv) > 1 else "127.0.0.1"
    recv_port = int(sys.argv[2]) if len(sys.argv) > 2 else 31001
    send_port = int(sys.argv[3]) if len(sys.argv) > 3 else 31002

    print("=== Lite3 NX Python Controller ===")
    print(f"Robot IP : {robot_ip}")
    print(f"Recv port: {recv_port} (SensorPacket from RK3588)")
    print(f"Send port: {send_port} (CommandPacket to RK3588)")
    print(f"Stand pose: {STAND_POSE}")

    # Run the dual-callback example by default.
    # Change to run_single_callback or run_sine_wave_example to test other modes.
    run_dual_callback(robot_ip, recv_port, send_port)


if __name__ == "__main__":
    main()
